numba/runtime/atomicops.py

```diff
@@ -0,0 +1,211 @@
+from collections import defaultdict
+
+from numba.config import MACHINE_BITS
+from numba import cgutils
+from llvmlite import ir, binding as llvm
+
+
+_word_type = ir.IntType(MACHINE_BITS)
+
+
+def _define_incref(module, atomic_incr):
+    """
+    Implement NRT_incref in the module
+    """
+    if "NRT_incref" not in module.globals:
+        return
+    fn_incref = module.get_global_variable_named("NRT_incref")
+    fn_incref.linkage = 'linkonce_odr'
+    builder = ir.IRBuilder(fn_incref.append_basic_block())
+    [ptr] = fn_incref.args
+    is_null = builder.icmp_unsigned("==", ptr, cgutils.get_null_value(ptr.type))
+    with cgutils.if_unlikely(builder, is_null):
+        builder.ret_void()
+    builder.call(atomic_incr, [builder.bitcast(ptr, atomic_incr.args[0].type)])
+    builder.ret_void()
+
+
+def _define_decref(module, atomic_decr):
+    """
+    Implement NRT_decref in the module
+    """
+    if "NRT_decref" not in module.globals:
+        return
+    fn_decref = module.get_global_variable_named("NRT_decref")
+    fn_decref.linkage = 'linkonce_odr'
+    calldtor = module.add_function(ir.FunctionType(ir.VoidType(),
+        [ir.IntType(8).as_pointer(), ir.IntType(32)]),
+        name="NRT_MemInfo_call_dtor")
+
+    builder = ir.IRBuilder(fn_decref.append_basic_block())
+    [ptr] = fn_decref.args
+    is_null = builder.icmp_unsigned("==", ptr, cgutils.get_null_value(ptr.type))
+    with cgutils.if_unlikely(builder, is_null):
+        builder.ret_void()
+    newrefct = builder.call(atomic_decr,
+                            [builder.bitcast(ptr, atomic_decr.args[0].type)])
+
+    refct_eq_0 = builder.icmp_unsigned("==", newrefct,
+                                       ir.Constant(newrefct.type, 0))
+    with cgutils.if_unlikely(builder, refct_eq_0):
+        do_defer = ir.Constant(ir.IntType(32), 0)
+        builder.call(calldtor, [ptr, do_defer])
+    builder.ret_void()
+
+
+def install_atomic_refct(module):
+    """
+    Implement both NRT_incref and NRT_decref in the module
+    """
+    incref = _define_atomic_inc_dec(module, "add", ordering='monotonic')
+    decref = _define_atomic_inc_dec(module, "sub", ordering='monotonic')
+
+    # Set LinkOnce ODR linkage
+    for fn in [incref, decref]:
+        fn.linkage = 'linkonce_odr'
+    del fn
+
+    _define_incref(module, incref)
+    _define_decref(module, decref)
+
+
+def _define_atomic_inc_dec(module, op, ordering):
+    """Define a llvm function for atomic increment/decrement to the given module
+    Argument ``op`` is the operation "add"/"sub".  Argument ``ordering`` is
+    the memory ordering.
+    """
+    ftype = ir.FunctionType(_word_type, [_word_type.as_pointer()])
+    fn_atomic = ir.Function(module, ftype, name="nrt_atomic_{0}".format(op))
+
+    [ptr] = fn_atomic.args
+    bb = fn_atomic.append_basic_block()
+    builder = ir.IRBuilder(bb)
+    ONE = ir.Constant(_word_type, 1)
+    builder.atomic_rmw(op, ptr, ONE, ordering=ordering)
+    res = builder.load(ptr)
```

Comment:
I think there is a race condition. If another thread runs in-between, then the two threads will see a resulting refcount of 0, so the destructor will be run twice. You want to reapply `op` to the result of the atomic_rmw instruction, instead (or you can change the decref test to check for 1, for example).